File: native/ios/ConverterCore/FormatTableSupport.swift
// Copyright (c) 2026 Owais Khan
// Licensed under the Apache License, Version 2.0

import Foundation

/// Matching logic for the generated format table.
///
/// The table itself is data compiled from `src/engine/formats.ts`; only this matching
/// behaviour is written by hand, and it is deliberately trivial so there is very little
/// room for it to disagree with the TypeScript reference implementation.
extension FormatTable {

    public static func spec(_ formatId: String) -> Spec? { specs[formatId] }

    /// True when any of the format's signatures matches the prefix.
    public static func matches(_ bytes: [UInt8], formatId: String) -> Bool {
        guard let spec = specs[formatId] else { return false }
        return spec.signatures.contains { matches(bytes, signature: $0) }
    }

    /// The note attached to whichever signature matched, for the detection reason.
    public static func note(_ bytes: [UInt8], formatId: String) -> String {
        guard let spec = specs[formatId] else { return "" }
        if let matched = spec.signatures.first(where: { matches(bytes, signature: $0) }) {
            return matched.note
        }
        return "\(spec.label) signature matched"
    }

    private static func matches(_ bytes: [UInt8], signature: Signature) -> Bool {
        signature.clauses.allSatisfy { clause in
            guard clause.offset >= 0, clause.offset + clause.bytes.count <= bytes.count else {
                return false
            }
            for (index, expected) in clause.bytes.enumerated()
            where bytes[clause.offset + index] != expected {
                return false
            }
            return true
        }
    }

    /// What a filename claims. Used only to report a mismatch, never to decide.
    public static func formatId(forFilename filename: String) -> String? {
        let ext = (filename as NSString).pathExtension.lowercased()
        guard !ext.isEmpty else { return nil }
        return allFormatIds.first { specs[$0]?.extensions.contains(ext) == true }
    }

    /// Every UTI the app registers as an "Open with" handler for.
    public static var allUTIs: [String] {
        allFormatIds.compactMap { specs[$0]?.uti }
    }
}

File: native/android/converter/FormatTableSupport.kt
// Copyright (c) 2026 Owais Khan
// Licensed under the Apache License, Version 2.0

package com.owaiskhan.converter.core

import com.owaiskhan.converter.format.FormatTable

/**
 * Matching logic for the generated format table.
 *
 * The table is data compiled from `src/engine/formats.ts`; only this matching is
 * written by hand, and it is deliberately trivial so there is very little room for it
 * to disagree with the TypeScript reference implementation or the Swift one.
 */
public object FormatMatcher {

  public fun spec(formatId: String): FormatTable.Spec? = FormatTable.SPECS[formatId]

  /** True when any of the format's signatures matches the prefix. */
  public fun matches(bytes: ByteArray, formatId: String): Boolean {
    val spec = FormatTable.SPECS[formatId] ?: return false
    return spec.signatures.any { matches(bytes, it) }
  }

  /** The note on whichever signature matched, used as the detection reason. */
  public fun note(bytes: ByteArray, formatId: String): String {
    val spec = FormatTable.SPECS[formatId] ?: return ""
    return spec.signatures.firstOrNull { matches(bytes, it) }?.note
      ?: "${spec.label} signature matched"
  }

  private fun matches(bytes: ByteArray, signature: FormatTable.Signature): Boolean =
    signature.clauses.all { clause ->
      if (clause.offset < 0 || clause.offset + clause.bytes.size > bytes.size) {
        false
      } else {
        clause.bytes.indices.all { i -> bytes[clause.offset + i] == clause.bytes[i] }
      }
    }

  /** What a filename claims. Used only to report a mismatch, never to decide. */
  public fun formatIdForFilename(filename: String): String? {
    val ext = filename.substringAfterLast('.', "").lowercase()
    if (ext.isEmpty() || ext == filename.lowercase()) return null
    return FormatTable.ALL_FORMAT_IDS.firstOrNull {
      FormatTable.SPECS[it]?.extensions?.contains(ext) == true
    }
  }

  /** Every MIME type the app registers intent filters for. */
  public val allMimeTypes: List<String>
    get() = FormatTable.ALL_FORMAT_IDS
      .mapNotNull { FormatTable.SPECS[it] }
      .flatMap { it.mimeTypes }
      .distinct()
}
